onmt/utils/Cuda.lua

```diff
@@ -52,16 +52,21 @@ end
   When using CPU only, converts to float instead of the default double.
 ]]
 function Cuda.convert(obj)
-  if torch.typename(obj) then
+  local objtype = torch.typename(obj)
+  if objtype then
     if Cuda.activated and obj.cuda ~= nil then
-      return obj:cuda()
+      if objtype:find('torch%..*LongTensor') then
+        return obj:cudaLong()
+      else
+        return obj:cuda()
+      end
     elseif not Cuda.activated and obj.float ~= nil then
       -- Defaults to float instead of double.
       return obj:float()
```

Comment:
Oh, we need to do the same way on the back to CPU:

```lua
if objtype:find('torch%..*LongTensor') then
  return obj:long()
else
  return obj:float()
end
```

With this final change we are good to go!